### ISSUE TITLE
Separate Ship and Player

### ISSUE THREAD

**Issue by robn** [pioneerspacesim/pioneer#38]
Right now `Ship` has too much knowledge and too many methods that are player-specific and should live in `Player`. Two examples:
- `SetNavTarget` & `SetCombatTarget` aren't required by NPCs; they're strictly UI functions. This is reinforced by the fact that these methods cause a sound to be played
- `AddMoney` and friends. NPC ships don't need money; they'll be equipped as necessary by ships. Indeed, `Player` should probably be a `MarketAgent`, but `Ship` should not.

Someone needs to go through and check and move things out of `Ship` into `Player` as appropriate. I think a good rule-of-thumb is that if it is a UI function in any way, then it belongs in `Player`. Another test might be to simply remove it from `Ship` and see if the code compiles.

**Comment by walterar** [comment]
It's a shame that after all this time no one has been interested in solving this. :(

**Comment by salborrelli** [comment]
@walterar FYI I've already "promised" to take a look at this one as soon as I have some time, see PR #3618 :smile:

**Comment by walterar** [comment]
Thank you very much Sal. All my thumb up be for you. :))

**Comment by salborrelli** [comment]
@robn @walterar after a quick look at the code I found that most (if not all) methods of the Ship class seems to be quite specific to the Ship, I haven't found any method which should logically be moved to the Player class.

Sure, some of the methods you mention in the above list (SetNavTarget, SetCombatTarget etc.) are part of the PlayerShipController class. Do you mean moving those methods from PlayerShipController to Player?

Another important thing to note is that, in the current version, the Player class inherits from Ship! Which means a Player doesn't really "owns" a Ship, because he "is" the Ship. Therefore, separating the Player from the Ship is not a matter of moving a few methods from here to there, we should go through a (possibly complex) re-ingegnerization of the code in order to separate the Player from the Ship, and to transform the Inheritance relationship to a Composition.

To summarize, given my far from complete knowledge of the game architecture, I'm probably missing something basic; that said, at the moment this issue is not completely clear to me. Can someone please describe the problem in more details please? Do you want to transform the Player-to-Ship relationship from Inheritance to Composition, or what?

**Comment by laarmen** [comment]
I might putting my fott in my mouth here, but my guess is that the
potential breakage will probably be at the Lua/C++ interface, and in our
Lua scripts. The compiler can tell us when the C++ side mixes the two
notions up, but not when it's in a mod.

**Comment by impaktor** [comment]
> Player class inherits from Ship! Which means a Player doesn't really "owns" a Ship, because he "is" the Ship. Therefore, separating the Player from the Ship is not a matter of moving a few methods from here to there, we should go through a (possibly complex) re-ingegnerization of the code in order to separate the Player from the Ship, and to transform the Inheritance relationship to a Composition.

That's probably hitting the nail smack on its head. Also, this issue probably involves more than renaming some functions, as it has been open for a few years, so not low hanging fruit, I guess.

> at the moment this issue is not completely clear to me. Can someone please describe the problem in more details please? Do you want to transform the Player-to-Ship relationship from Inheritance to Composition, or what?

(Part of) the goal here is to allow a player to _have_ a ship, so player should not inherit from the ship class, so restructure might go deep. I'm not familiar with this code myself, and I'm quite pressed for time at the moment.

Worth keeping in mind is that perhaps in the future we could even allow the player to not have a ship at all, but just a space suit (that would essentially "be a ship", perhaps?). At least that is one way I imagine a possible outcome of this code restructure.

Alas, only @fluffyfreak remains of the c++ knowledgeable people, as old man @robn has retired. (let's see if that pissed him off enough to post something that might enlighten us).

I wish I could help more, and sorry if I sent you down a rabbit hole here. Don't feel pressured to continue.

PS. _having been beaten by our LUA/C++ expert @laarmen, in replying, I'm starting to suspect that if this change is not isolated to C++, the rabbit hole might just have gotten deeper._

**Comment by walterar** [comment]
@salborrelli My immediate need is to have a way in LUA to cancel or deselect SetCombatTarget, or SetNavTarget.
It may be simply that Game.player:SetCombatTarget() deselect, instead of generating an error. That is, that C ++ understands that should put SetCombatTarget(0)
Here's a way this happens in Player.cpp when it comes into hyperspace.
```
void Player::OnEnterHyperspace()
{
	s_soundHyperdrive.Play("Hyperdrive_Jump");
	SetNavTarget(0);
	SetCombatTarget(0);
```

**Comment by fluffyfreak** [comment]
There you go @walterar, PR #3634, please test :)

**Comment by fluffyfreak** [comment]
@salborrelli I think that in English we'd commonly say `refactor`, I'm not familiar with `re-ingegnerization` but I guess that you mean something similar.

Yes, the idea is that a ship would be own via composition rather than being a ship via inheritance. NPC/AI might also need to own a ship instead of being one but it depends how you would like to rewrite it :)

Also as @impaktor says this is potentially quite a large task so don't feel that you must work on it.

> Alas, only @fluffyfreak remains of the c++ knowledgeable people

What! Oh damn when did that happen! @johnbartholomew is still doing stuff too I'm sure.

**Comment by salborrelli** [comment]
@fluffyfreak with apologies for my bad english, yes I meant "re-engineering" which sounds like a kind of re-factoring infact.

>Also as @impaktor says this is potentially quite a large task so don't feel that you must work on it.

Agree, but if you think I can help someway, let me know ...

>Alas, only @fluffyfreak remains of the c++ knowledgeable people

Actually I know zero about LUA and am not that good in graphic design, so I guess that C++ is the only part of this project I can be of any help on :smile:

**Comment by mike-f1** [comment]
I all,
I take a glance on this time ago and now I was back
trying to help:
https://github.com/pioneerspacesim/pioneer/issues/2012

but facing some trouble with AI inheritance I start searching
other related issues and I found this one and also this:
https://github.com/pioneerspacesim/pioneer/issues/22

...so I think that if you would move in the right direction
you have to take into consideration also these issues.

**Comment by impaktor** [comment]
@mike-f1 nice! Do what you can. But be aware that this issue is a very big undertaking. Perhaps you want to "warm up" with something smaller. Either way, looking forward to see what you come up with!

**Comment by mike-f1** [comment]
I did something for Missile class trying to inherit from DynBody but near finish I face some
problem with AI methods (Missile need to land?), so my little plan stops.

Here my great plan for https://github.com/pioneerspacesim/pioneer/issues/2012 :
A) Create a class holding "propulsion" and tank (copying mostly methods from ship)
B) Create classes for flying AI (basic methods and some split for AI purpose, mainly for
     docking and hyperspace capability)
C) Redo class Missile from DynBody and these new classes
D) Redo Ship class as a inheritance with DynBody and these new classes

But as you can see this plan probably will modify code in deep and I think has to be
coupled with the above mentioned issues, so I need more info from the team.

PS: Is this the best place to discuss these changes?

**Comment by impaktor** [comment]
I don't know the code well enough to be of much help here, sorry.

> PS: Is this the best place to discuss these changes?

IRC, preferably evening, GMT.

**Comment by mike-f1** [comment]
I did some code, isn't working because in the mean I obviously forget
something (you cannot rotate ships...), but I think it can help sharing
the idea I have on how it could work:
https://github.com/mike-f1/pioneer/tree/modular_ship

The final goal is to do something like this also for AI and then you can
"modularize" things... Any comments & feedbacks & critics is appreciated...

**Comment by mike-f1** [comment]
Here an update: https://github.com/mike-f1/pioneer/tree/modular_ship

It is not working entirely and it needs to be tuned, but this is my approach to the problem

@fluffyfreak , @robn , @ecraven : could you put a glance here and say me if my approach
is completely wrong or if is acceptable?

Thank's
